src/main/java/com/minecolonies/blocks/BlockBarrel.java

```diff
@@ -0,0 +1,216 @@
+    /**
+     * The resistance this block has.
+     */
+    private static final float                      RESISTANCE     = 1F;
+    private static final HashMap<BlockPos, Integer> timers         = new HashMap<>();
+    private static final HashMap<BlockPos, Integer> fillings       = new HashMap<>();
+
+    public BlockBarrel()
+    {
+        super(Material.wood);
+        this.setDefaultState(this.blockState.getBaseState().withProperty(BARRELSTATE, BARRELSTATE_FILLING));
+
+        this.setTickRandomly(true);
+        initBlock();
+    }
+
+    /**
+     * initialize the block
+     */
+    private void initBlock()
+    {
+        setRegistryName(BLOCK_NAME);
+        setUnlocalizedName(String.format("%s.%s", Constants.MOD_ID.toLowerCase(), BLOCK_NAME));
+        setCreativeTab(ModCreativeTabs.MINECOLONIES);
+        GameRegistry.registerBlock(this, BLOCK_NAME);
+        setHardness(BLOCK_HARDNESS);
+        setResistance(RESISTANCE);
+    }
+
+    public void AddItemToBarrel(World worldIn, EntityPlayer playerIn, ItemStack itemStack, IBlockState state, BlockPos pos)
+    {
+        UseBarrel(worldIn, playerIn, itemStack, state, pos);
+    }
+
+    //whenever player right click to barrel call this.
+    public boolean UseBarrel(World worldIn, EntityPlayer playerIn, ItemStack itemstack, IBlockState state, BlockPos pos)
+    {
+        Log.logger.info("block activated");
+
+        int barrelState = state.getValue(BARRELSTATE);
+        int fullness = fillings.getOrDefault(pos, 0);
+
+        Log.logger.info("At this moment bs= " + barrelState + " and fl=" + fullness);
+
+
+        //if statement 1
+        if (state.getValue(BARRELSTATE) == BARRELSTATE_DONE)
+        {
+            // todo: add this back in once compost exists again
+            // playerIn.inventory.addItemStackToInventory(new ItemStack(ModItems.compost, 8));
+            worldIn.setBlockState(pos, state.withProperty(BARRELSTATE, BARRELSTATE_FILLING));
+            fillings.put(pos, MIN_FULLNESS);
+            Log.logger.info("Set Blockstate to " + worldIn.getBlockState(pos));
+            return true;
+        }
+
+        if (itemstack == null)
+        {
+            return true;
+        }
+
+        Item item = itemstack.getItem();
+
+        if (item == Items.rotten_flesh && barrelState == BARRELSTATE_FILLING)
+        {
+            Log.logger.info("item Consumed");
+
+            itemstack.stackSize--;
+
+            fullness += 1;
+            if (fullness >= MAX_FULLNESS)
+            {
+                fullness = MAX_FULLNESS;
+                worldIn.setBlockState(pos, state.withProperty(BARRELSTATE, BARRELSTATE_COMPOSTING));
+            }
+            fillings.put(pos, fullness);
+            Log.logger.info("now FULLNESS = " + fullness);
+
+            return true;
+        }
+
+        return true;
+    }
+
+    public void GetItemFromBarrel(World worldIn, EntityPlayer playerIn, ItemStack itemStack, IBlockState state, BlockPos pos)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "itemStack". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1172)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this method name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00100)